src/parallel.cpp

```diff
@@ -17,38 +17,49 @@
 
 #include <boost/thread/thread.hpp>
 
-#define MAX_SCRIPT_CHECK_QUEUES 4
-
-static CCheckQueue<CScriptCheck> scriptcheckqueue1(128);
-static CCheckQueue<CScriptCheck> scriptcheckqueue2(128);
-static CCheckQueue<CScriptCheck> scriptcheckqueue3(128);
-static CCheckQueue<CScriptCheck> scriptcheckqueue4(128);
-
 using namespace std;
 
+static const unsigned int nScriptCheckQueues = 4;
+std::unique_ptr<CParallelValidation> PV;
 
-void AddScriptCheckThreads(int i, CCheckQueue<CScriptCheck> *pqueue)
+static void HandleBlockMessageThread(CNode *pfrom, const std::string strCommand, const CBlock block, const CInv inv);
+
+static void AddScriptCheckThreads(int i, CCheckQueue<CScriptCheck> *pqueue)
 {
     ostringstream tName;
     tName << "bitcoin-scriptchk" << i;
     RenameThread(tName.str().c_str());
     pqueue->Thread();
 }
 
-void AddAllScriptCheckQueuesAndThreads(int nScriptCheckThreads, boost::thread_group *threadGroup)
+CParallelValidation::CParallelValidation(int threadCount, boost::thread_group *threadGroup)
+    : semaphore(nScriptCheckQueues)
 {
-    CCheckQueue<CScriptCheck> *pqueue[MAX_SCRIPT_CHECK_QUEUES] = {
-        &scriptcheckqueue1, &scriptcheckqueue2, &scriptcheckqueue3, &scriptcheckqueue4};
+    if (threadCount <= 1)
+        threadCount = 0;
+    else if (threadCount > MAX_SCRIPTCHECK_THREADS)
+        threadCount = MAX_SCRIPTCHECK_THREADS;
+    nThreads = threadCount;
+
+    LogPrintf("Using %d threads for script verification\n", threadCount);
 
-    for (int i = 0; i < MAX_SCRIPT_CHECK_QUEUES; i++)
+    while (QueueCount() < nScriptCheckQueues)
     {
-        allScriptCheckQueues.Add(pqueue[i]);
-        for (int j = 0; j < nScriptCheckThreads; j++)
-            threadGroup->create_thread(boost::bind(&AddScriptCheckThreads, i + 1, pqueue[i]));
+        std::unique_ptr<CCheckQueue<CScriptCheck> > queue(new CCheckQueue<CScriptCheck>(128));
+
+        for (unsigned int i = 0; i < nThreads; i++)
+            threadGroup->create_thread(boost::bind(&AddScriptCheckThreads, i + 1, queue.get()));
+
+        vQueues.push_back(std::move(queue));
```

Comment:
Also, I just noticed that nScriptCheckQueues is a static const int of 4.  So this whole conversion does not add functionality (I originally thought it added the ability to have a variable number of script check queues).

So this change takes some globally defined objects that are supposed to persist through the entire program's life and converts them to dynamically allocated objects that may be deleted.  This adds confusion and complexity.  

Although "reducing globals" is often stated as a goal, saying that is a summary for the real purpose.  The real purpose of reducing globals is to carefully control access to subsystems through well defined APIs rather than allowing any section of code to directly access and modify the state of your subsystem.  In other words, reduce global scope rather than reduce globals.   So instead of allocating stuff on the heap, this can be achieved by putting the global inside another object (perhaps the PV object?)  Or by putting it under a namespace or statically defining it in a module.

Also, I do not think that a refactor just to c++-ify code is a good use of our time.  Generally the best refactorizations reorganize code to add clarity across the rest of the code base.   It you disagree, perhaps we can discuss this as a group next tuesday.